src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,119 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+
+    private static final int defaultSize = 10;
+    private Object[] arrayList = new Object[defaultSize];
+    private int size = 0;
+
+    private void grow() {
+        arrayList = Arrays.copyOf(arrayList, (int)(arrayList.length * 1.5));
+    }
+
     @Override
     public void add(T value) {
-
+        if (arrayList.length == size()) {
+            grow();
+        }
+        arrayList[size()] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size()) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index + " out of bounds");
+        }
+        if (size() == arrayList.length) {
+            grow();
+        }
+        Object[] newList = new Object[arrayList.length];
 
+        for (int i = 0; i < index; i++) {
+            newList[i] = arrayList[i];
+        }
+        newList[index] = value;
+        for (int i = index; i < size(); i++) {
+            newList[i + 1] = arrayList[i];
+        }
+        arrayList = newList;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        if (arrayList.length < size() + list.size()) {
+            grow();
+        }
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index >= size() || index < 0) {
```

Comment:
The index check logic is repeated in multiple methods (`get`, `set`, `remove`). Consider extracting this logic into a separate private method to avoid code duplication.